registry/storage/purgeuploads.go

```diff
@@ -0,0 +1,144 @@
+package storage
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	storageDriver "github.com/docker/distribution/registry/storage/driver"
+)
+
+// TODO: logrus
+
+// uploadData defines the temporary files created during a layer upload
+type uploadData struct {
+	dataFilePath string
+	dateFilePath string
+	startedAt    time.Time
+}
+
+// TODO(richard): return containing folder
+func (ud *uploadData) Files() []string {
+	files := []string{ud.dataFilePath}
+	return append(files, ud.dateFilePath)
+}
+
+// PurgeUploads deletes files from the upload directory
+// created before olderThan.  The list of files deleted and errors
+// encountered are returned
+func PurgeUploads(driver storageDriver.StorageDriver, olderThan time.Time, actuallyDelete bool) ([]string, []error) {
+	// TODO(richard): normalize error type
+	uploadData, errors := getOutstandingUploads(driver)
+	var deleted []string
+	for _, uploadData := range uploadData {
+		for _, file := range uploadData.Files() {
+			if uploadData.startedAt.Before(olderThan) {
+				var err error
+				if actuallyDelete {
+					err = driver.Delete(file)
+				}
+				if err == nil {
+					deleted = append(deleted, file)
+				} else {
+					errors = append(errors, err)
+				}
+			}
+		}
+	}
+	return deleted, errors
+}
+
+// getOutstandingUploads walks the upload directory, collecting files
+// which could be eligible for deletion
+func getOutstandingUploads(driver storageDriver.StorageDriver) (map[string]uploadData, []error) {
+	var errors []error
+	outstandingUploads := make(map[string]uploadData, 0)
+	pm := defaultPathMapper
+
+	pathComponentRegexp := `([a-z0-9._-]+)+` // TODO(richard): stronger regex for uuid?
+	uploadFileRegexStr, err := pm.path(
+		uploadDataPathSpec{name: pathComponentRegexp, uuid: pathComponentRegexp})
+	if err != nil {
+		return outstandingUploads, pushError(errors, "", err)
+	}
+	uploadFileRegex := regexp.MustCompile(uploadFileRegexStr)
+
+	uploadDateRegexStr, err := pm.path(
+		uploadStartedAtPathSpec{name: pathComponentRegexp, uuid: pathComponentRegexp})
+	if err != nil {
+		return outstandingUploads, pushError(errors, "", err)
+	}
+	uploadDateRegex := regexp.MustCompile(uploadDateRegexStr)
+
+	// TODO(richard): path from pm
+	root := "/docker/registry/v2/repositories"
+	err = Walk(driver, root, func(fileInfo storageDriver.FileInfo) error {
+		path := fileInfo.Path()
+		uuid, err := match(path, uploadFileRegex)
+		if err != nil {
+			errors = pushError(errors, path, err)
+			goto nextFile
```

Comment:
may be I was not clear enough =). I meant that you can uese ```return nil``` here instead of ```goto```. ```goto``` is good if you are gonna to add more actions except ```return nil``` like logging or anything else to prevent code duplication.